// @progress/kendo-schematics/src/collection/grid/files/inline/__name@dasherize__/__name@dasherize__.component.ts
import { Component } from '@angular/core';<% if (dataPath) {%>
import { <%= dataName %> } from '<%= dataPath %>'; <% } %><% if (mockedData) {%>
import { products } from './data'; <% } %>

@Component({
	selector: '<%= dasherize(name) %>-component',
	template: `
		<kendo-grid 
			<% if (paging || sorting || filtering || grouping) {%>[kendoGridBinding]="gridData"<% } else {%>[data]="gridData"<% } %>
			[filterable]="<% if (filtering) {%> true<% } else {%>false<% } %>"
			[groupable]="<% if (grouping) {%>true<% } else {%>false<% } %>"
			[pageable]="<% if (paging) {%>true<% } else {%>false<% } %>"
			[pageSize]="pageSize"
			[sortable]="<% if (sorting) {%>true<% } else {%>false<% } %>"
			<% if (height) {%>[height]="<%=height%>"<% } %>><% if (columns.length > 0) {%>
			<kendo-grid-column
				*ngFor="let field of fields"
				[field]="field">
          	</kendo-grid-column> <% } %>
		</kendo-grid>
	`,<% if (inlineStyle) {%>
	styles: [``]<% } else {%>
	styleUrls: ['./<%= dasherize(name) %>.component.css']<% } %>
})
export class <%= classify(name) %>Component  {
	public gridData: any[] = <% if (dataName) {%><%=dataName%><% } else {%>products<% }%>;<% if (columns.length > 0) {%>
	public fields = [<%=arrayify(columns)%>];<% }%>
    public pageSize: number = <%=pageSize%>;
}
